### ISSUE TITLE
Iteration closing with `throw` on delegated yield

### ISSUE THREAD

**Issue by zloirock** [facebook/regenerator#185]
[14.4.14 Runtime Semantics: Evaluation -> YieldExpression : yield * AssignmentExpression -> 6 b iii - iv](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-generator-function-definitions-runtime-semantics-evaluation)
> If iterator does not have a throw method, this throw is going to terminate the yield* loop. But first we need to give iterator a chance to clean up.

[Demo](http://goo.gl/cgpwd9):
```js
var iter = [null, null, null].values();
//iter.throw = function(){
//  console.log('throw');
//  return {done: true};
//};
iter.return = function(){
  console.log('return');
  return {done: true};
}
var gen = (function*(){
  try {
    yield *iter;
  } catch(e){}
})();
gen.next();
gen.throw();
```
Expected: 'return'
Result: nothing

**Comment by zloirock** [comment]
@benjamn ping

**Comment by zloirock** [comment]
Found  similar in babel repo https://github.com/babel/babel/issues/1046

**Comment by benjamn** [comment]
Great catch @zloirock. In fact, you discovered an area of the spec where neither Node v0.11 nor Node v0.12 get things completely right. Please check out the tests included with my commit to verify that my implementation matches your understanding of the spec!

**Comment by zloirock** [comment]
Thanks :)

**Comment by rauschma** [comment]
Have you considered this related case?

```js
function* callee() {
    try {
        yield 'b';
        yield 'c';
    } finally {
        console.log('finally callee');
    }
}
function* caller() {
    try {
        yield 'a';
        yield* callee();
        yield 'd';
    } finally {
        console.log('finally caller');
    }
}
let generatorObject = caller();
console.log(generatorObject.next().value); // 'a'
console.log(generatorObject.next().value); // 'b'
generatorObject.return();
```
The output is:

```
a
b
finally callee
```

The output should be:

```
a
b
finally callee
finally caller
```

**Comment by benjamn** [comment]
@rauschma what happens if the delegate iterator throws an exception in response to calling its `.return` method? Should that exception be ignored, or should it be thrown from the `yield*` expression in the caller?

**Comment by benjamn** [comment]
Fixed: 99f0d26f9f2e4e217b8e9c014f69e9df373c3b64

**Comment by rauschma** [comment]
@benjamn:
> what happens if the delegate iterator throws an exception in response to calling its .return method? Should that exception be ignored, or should it be thrown from the yield* expression in the caller?

It should be thrown from the `yield*` expression, AFAICT.
